### PR TITLE
[mono] Add suport for denied assemblies, fixed bxc#56682

### PR DESCRIPTION
**Pull request by radical**
Mono denies loading certain "bad" assemblies because they include an
implementation that doesn't work under mono. Instead, mono provides it's
own implementations of the same which are safe to load.

These are specific versions from specific nugets. Mono installs a list
of these in $(MSBuildExtensionsPath)/deniedAssembliesList.txt .

If these bad assemblies get bundled with an app and the runtime tries to
load them, then they will get denied. If this is running from a full
mono installation, then the runtime continues looking for the assembly
in other locations.

But if this is a case, like a Mac app bundle (with Xamarin.Mac), which
uses it's own bundled mono+assemblies, then the runtime won't be able to
fallback to anything else. So, we want to ensure that we bundle the
"good" assemblies instead.

This is not an issue for XA/XI as they have their own
TargetFrameworkMoniker and so the nuget has `_._` for them. IOW, msbuild
uses the framework assemblies there.

Solution:

- We tried an approach of: filtering out such references
  (`@(Reference)`) before the `ResolveAssemblyReference` task. All the
  facades are passed to the compiler as references, so the correct
  assemblies will get pulled in accordingly.

  But the problem with this was that if after removing these, we ended
  up with no facade references, then `ResolveAssemblyReference` returns
  `DependsOnSystemRuntime=false`, which causes the facade assemblies to
  not be implicitly passed to the compiler!

- Instead, now we get the list of target framework directories and
  facade directories and look for the "good" assemblies in there. If we
  can find the assembly, then we update the reference (ItemSpec or HintPath) to
  the new path.
  Else, we warn the user and don't touch the reference!

For this, we add a `FilterDeniedAssemblies` tasks in a new
`Mono.Build.Tasks.dll`.

Note: the list is part of this commit for now. Before the merge, this
will be generated+installed from mono instead.

### CODE REVIEW

**Comment by radical** [comment]
This is not the final thing! The list of denied assembly is assumed to have been installed by mono in `lib/mono/xbuild`. Since, mono doesn't do that yet, the file is bundled in the commit and copied by hand.

Note: the list is part of this commit for now. Before the merge, this will be generated+installed from mono instead.

**Comment by radical** [comment]
Oh and this needs *testing*!! :)

File: mono/tasks/FilterDeniedAssemblies.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        using (var stream = File.OpenRead(fullpath))
+        using (var peFile = new PEReader(stream))
+        {
+            var metadataReader = peFile.GetMetadataReader();
+
+            var entry = metadataReader.GetAssemblyDefinition();
+            ver = entry.Version;
+            var guid = metadataReader.GetGuid(metadataReader.GetModuleDefinition().Mvid);
+
+            var id = guid.ToString ().ToUpper  ();
```
Comment (marek-safar) [inline mono/tasks/FilterDeniedAssemblies.cs]: Should be ToUpperInvariant and ToString with culture.invariantculture

File: mono/ExtensionsPath-ToolsVersion/Microsoft.Common.targets/ImportAfter/Microsoft.Common.Mono.After.targets [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+    <UsingTask TaskName="FilterDeniedAssemblies" AssemblyFile="$(MSBuildExtensionsPath)\Mono.Build.Tasks.dll" />
+
+    <Target Name="FilterDeniedAssemblyReferences" BeforeTargets="ResolveAssemblyReferences" Condition="'$(SkipFilterBadReferences)' != 'true'">
+        <FilterDeniedAssemblies References="@(Reference)">
+            <Output TaskParameter="DeniedReferences" ItemName="DeniedReferences" />
+            <Output TaskParameter="FilteredReferences" ItemName="FilteredReferences" />
+        </FilterDeniedAssemblies>
+
+        <Message Text="Removed references to assemblies that are denied loading by Mono runtime: %(DeniedReferences.DeniedAssemblyPath)" Importance="Low" />
```
Comment (chamons) [inline mono/ExtensionsPath-ToolsVersion/Microsoft.Common.targets/ImportAfter/Microsoft.Common.Mono.After.targets]: As a user, I have no clue what this means. Either we should remove the message, have it do something like I did in the IDE:

```DebuggerLoggingService.LogAndShowException ("Unable to launch application to debug, possibly due to it crashing quickly.\nSee https://aka.ms/xm-crash-no-output for more help.", e);```

or rephrase it to something like:

"Ignoring references to assembles {} as they are denied loading by Mono runtime". That way it is at least clear that we saw them and are doing nothing, vs remove, which suggests change (did they change my project? do I need to do anything?).

File: mono/tasks/FilterDeniedAssemblies.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+using System.Linq;
+
+public class FilterDeniedAssemblies : Task
+{
+    static Lazy<ExclusionDB> db = new Lazy<ExclusionDB>();
+
+    public override bool Execute ()
+    {
+        if (db.Value.Empty) {
+            Log.LogMessage(MessageImportance.Low, "No list of denied assemblies was found, so not filtering any references.");
```
Comment (chamons) [inline mono/tasks/FilterDeniedAssemblies.cs]: Is this something a user should see? I'd suggest either direct them to file a bug at bugzilla.xamarin.com/index.cgi or make it debug only?

File: mono/tasks/FilterDeniedAssemblies.cs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        var filename = Path.GetFileName (fullpath);
+        Version ver;
+
+        using (var stream = File.OpenRead(fullpath))
+        using (var peFile = new PEReader(stream))
+        {
+            var metadataReader = peFile.GetMetadataReader();
+
+            var entry = metadataReader.GetAssemblyDefinition();
+            ver = entry.Version;
```
Comment (chamons) [inline mono/tasks/FilterDeniedAssemblies.cs]: Can metadataReader.GetAssemblyDefinition() ever return null?

**Review by chamons — APPROVED** [review]
I'd love a test (maybe in xamarin-macios) for this somewhere.

File: mono/tasks/FilterDeniedAssemblies.cs (hunk 3 above)
Comment (radical) [reply mono/tasks/FilterDeniedAssemblies.cs]: This would be seen only in a detailed or diagnostic log. Suggesting them to file a bug sounds like a good idea.

File: mono/tasks/FilterDeniedAssemblies.cs (hunk 4 above)
Comment (radical) [reply mono/tasks/FilterDeniedAssemblies.cs]: The documentation wasn't very helpful. Source we'll have a non-null value or an exception. https://github.com/dotnet/corefx/blob/master/src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.cs#L1048-L1055

File: mono/tasks/FilterDeniedAssemblies.cs (hunk 3 above)
Comment (chamons) [reply mono/tasks/FilterDeniedAssemblies.cs]: We have some messages like that in mmp:

https://github.com/xamarin/xamarin-macios/blob/master/tools/mmp/driver.cs#L846

File: mono/ExtensionsPath-ToolsVersion/Microsoft.Common.targets/ImportAfter/Microsoft.Common.Mono.After.targets (hunk 2 above)
Comment (radical) [reply mono/ExtensionsPath-ToolsVersion/Microsoft.Common.targets/ImportAfter/Microsoft.Common.Mono.After.targets]: Sure, your message makes sense. This is also a detailed/diagnostic message, and completely normal, doesn't indicate any issues, so I don't see why it should be surfaced in the IDE. Feel free to correct me of course :)

**Comment by radical** [comment]
re:test, oh, I am not declaring this done yet! I will create a wrench lane to get a build for this and then you can test it. I tried it basically with your `TestRoslynApp` sample and one other project, essentially.

**Comment by radical** [comment]
re:test, um but yeah, we should be adding a test for this, XM is a good place for it.

**Comment by radical** [comment]
Updated with a different approach.

File: install-mono-prefix.sh [hunk 5]
```diff
@@ -77,6 +77,10 @@ test -d ${DESTDIR}${XBUILD_DIR}/14.0/Imports && cp -R ${DESTDIR}${XBUILD_DIR}/14
 cp -R nuget-support/tv/* ${DESTDIR}${XBUILD_DIR}/$MSBUILD_TOOLSVERSION
 cp -R nuget-support/tasks-targets/* ${DESTDIR}${XBUILD_DIR}/
 
+#
+mv ${MSBUILD_OUT_DIR}/Mono.Build.Tasks.* ${DESTDIR}${XBUILD_DIR}
```
Comment (chamons) [inline install-mono-prefix.sh]: Should this be copy?

File: mono/tasks/FilterDeniedAssemblies.cs [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+using System.Linq;
+using System.Text;
+
+namespace Mono.Build.Tasks
+{
+    public class FilterDeniedAssemblies : Task
+    {
+        static string s_deniedListFilename = "deniedAssembliesList.txt";
```
Comment (chamons) [inline mono/tasks/FilterDeniedAssemblies.cs:17]: I've not seeing type prefixes in variables names before? Is this an msbuild ism?

Comment (radical) [reply mono/tasks/FilterDeniedAssemblies.cs:17]: This is just for *static* fields, and yeah it's a msbuild code style (which made things a little easier to read here too!). The source file has a mixed mono and msbuild style though, which I should clean up too!

File: install-mono-prefix.sh (hunk 5 above)
Comment (radical) [reply install-mono-prefix.sh]: Yeah :)

File: mono/tasks/FilterDeniedAssemblies.cs [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        }
+
+        /// <summary>
+        /// Stringify a list of strings, like {"abc, "def", "foo"} to "abc, def and foo"
+        /// or {"abc"} to "abc"
+        /// <param name="strings">List of strings to stringify</param>
+        /// <returns>Stringified list</returns>
+        /// </summary>
+        static string StringifyList(string[] strings)
+        {
```
Comment (chamons) [inline mono/tasks/FilterDeniedAssemblies.cs]: Not to play C# golf but...


```		
        static string Stringify (List<string> list)
		{
			var spaceSeperated = String.Join (", ", list);
			if (list.Count > 1)
				spaceSeperated = spaceSeperated.Insert (spaceSeperated.LastIndexOf (',') + 1, " and");
			return spaceSeperated;
		}
```

**Comment by radical** [comment]
Updated to reflect @chamons' suggested changes.

File: mono/ExtensionsPath-ToolsVersion/Microsoft.Common.targets/ImportAfter/Microsoft.Common.Mono.After.targets [hunk 8]
```diff
@@ -0,0 +1,32 @@
+﻿<!--
+***********************************************************************************************
+Microsoft.Common.Mono.targets
```
Comment (akoeplinger) [inline mono/ExtensionsPath-ToolsVersion/Microsoft.Common.targets/ImportAfter/Microsoft.Common.Mono.After.targets]: should be `Microsoft.Common.Mono.After.targets` to match the file name.

File: mono/tasks/FilterDeniedAssemblies.cs [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+        public override bool Execute ()
+        {
+            try {
+                if (s_db.Value.Empty) {
+                    // nothing to filter!
+                    FilteredReferences = References;
+
+                    return !Log.HasLoggedErrors;
+                }
+            } catch (FileNotFoundException fe) when (fe.FileName != null && String.Compare(Path.GetFileName(fe.FileName), s_deniedListFilename) == 0) {
```
Comment (akoeplinger) [inline mono/tasks/FilterDeniedAssemblies.cs]: I've been told by Zoltan that exception filter clauses should be avoided if possible. Is there a reason why we need to catch FileNotFoundException instead of simply doing `if (!File.Exists())`?

File: mono/tasks/FilterDeniedAssemblies.cs [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+                        else
+                            referenceItem.ItemSpec = safeAssemblyFilePath;
+
+                        Log.LogMessage (MessageImportance.Low, $"Changed the denied assembly reference path from {assemblyPathFromReference} to the safe assembly path {safeAssemblyFilePath}.");
+                    } else {
+                        // Warn and don't touch the reference
+                        Log.LogWarning(null, "MSB3912", null, null, -1, -1, -1, -1,
+                                        $"INTERNAL WARNING: Could not find the replacement assembly ({assemblyFilename}) for the denied reference {assemblyPathFromReference}" +
+                                        $" in the search paths {StringifyList(SearchPaths)}. This might cause issues at runtime." + 
+                                        " Please file a bug report at http://bugzilla.xamarin.com .");
```
Comment (akoeplinger) [inline mono/tasks/FilterDeniedAssemblies.cs]: https please

File: mono/tasks/FilterDeniedAssemblies.cs [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+                    // nothing to filter!
+                    FilteredReferences = References;
+
+                    return !Log.HasLoggedErrors;
+                }
+            } catch (FileNotFoundException fe) when (fe.FileName != null && String.Compare(Path.GetFileName(fe.FileName), s_deniedListFilename) == 0) {
+                // If the exception is about the missing denied list, then just warn.
+                // Any other errors can surface
+                if (!s_haveWarnedAboutMissingList) {
+                    Log.LogWarning(null, "MSB3911", null, null, -1, -1, -1, -1, $"INTERNAL WARNING: {fe.Message}. Please file a bug report at http://bugzilla.xamarin.com .");
```
Comment (akoeplinger) [inline mono/tasks/FilterDeniedAssemblies.cs]: https please

File: mono/tasks/FilterDeniedAssemblies.cs [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+
+        /// <summary>
+        /// Stringify a list of strings, like {"abc, "def", "foo"} to "abc, def and foo"
+        /// or {"abc"} to "abc"
+        /// <param name="strings">List of strings to stringify</param>
+        /// <returns>Stringified list</returns>
+        /// </summary>
+        static string StringifyList(string[] strings)
+        {
+            if (strings?.Length == 0)
```
Comment (akoeplinger) [inline mono/tasks/FilterDeniedAssemblies.cs]: this won't catch a null `strings` and make String.Join() crash later on.

File: mono/tasks/FilterDeniedAssemblies.cs [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+                return String.Empty;
+
+            var commaSeparatedString = String.Join(", ", strings);
+            if (strings.Length > 1)
+            {
+                commaSeparatedString = commaSeparatedString.Insert(commaSeparatedString.IndexOf(",") + 1, " and");
```
Comment (akoeplinger) [inline mono/tasks/FilterDeniedAssemblies.cs]: this would print `"abc, def, and foo"` (note comma before `and`).

What about:

```csharp
if (strings.Length == 1)
   return strings[0];

var allButLast = new ArraySegment<string>(strings, 0, strings.Length-1);
return String.Join(",", allButLast) + " and " + strings[strings.Length-1];
```

Though I wonder whether we need to get this fancy at all in a simple warning message? I think a simple `String.Join(",", strings)` should do it as well.

**Review by akoeplinger — CHANGES_REQUESTED** [review]

File: mono/tasks/FilterDeniedAssemblies.cs (hunk 13 above)
Comment (radical) [reply mono/tasks/FilterDeniedAssemblies.cs]: I have an updated patch which I haven't pushed yet. And that removes this method as I am not including the list of search paths in the user-visible warning. It didn't seem to have any value for a user. It would be useful only for somebody debugging, who will get the list in a diagnostic log.

File: mono/tasks/FilterDeniedAssemblies.cs (hunk 9 above)
Comment (radical) [reply mono/tasks/FilterDeniedAssemblies.cs]: Does it not work well on mono? I can rework to avoid it.

File: mono/tasks/FilterDeniedAssemblies.cs (hunk 12 above)
Comment (radical) [reply mono/tasks/FilterDeniedAssemblies.cs]: yikes! thanks!

File: mono/tasks/FilterDeniedAssemblies.cs (hunk 9 above)
Comment (akoeplinger) [reply mono/tasks/FilterDeniedAssemblies.cs]: Apparently it has some issues, can't remember the exact details. Better to rework it if it's easy :)

**Comment by radical** [comment]
Updated. Hopefully, this is ready this time :) Added a test project too.

**Comment by radical** [comment]
Updated again: made this an opt-in feature, via `$(FilterDeniedAssemblies)`, just to reduce the impact of this. We can enable this in XM for now. And any other profile owners that we might have, can do the same.

**Review by akoeplinger — APPROVED** [review]

**Comment by borgdylan** [comment]
Will the latest SDK bits be added as well? At present NETStandard 2.0 support is completely missing.

**Comment by chamons** [comment]
@borgdylan https://gitter.im/xamarin/xamarin-macios (or the other associated channels) is a more appropriate location for such questions.

Though if you are asking about Xamarin.Mac or iOS - https://medium.com/@donblas/xamarin-mac-and-netstandard2-708a06890302

**Comment by borgdylan** [comment]
I had asked under the msbuild gitter (was asking about mono msbuild fork) and got no response.
